source/stun_serializer.c

```diff
@@ -0,0 +1,897 @@
+
+    if( ( pCtx == NULL ) ||
+        ( pAttributeValueBuffer == NULL ) ||
+        ( attributeValueBufferLength == 0 ) )
+    {
+        result = STUN_RESULT_BAD_PARAM;
+    }
+
+    if( ( result == STUN_RESULT_OK ) &&
+        ( pCtx->pStart != NULL ) )
+    {
+        if( STUN_REMAINING_LENGTH( pCtx ) < STUN_ATTRIBUTE_TOTAL_LENGTH( attributeValueLengthPadded ) )
+        {
+            result = STUN_RESULT_OUT_OF_MEMORY;
+        }
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        if( ( pCtx->attributeFlag & STUN_FLAG_FINGERPRINT_ATTRIBUTE ) != 0 )
+        {
+            /* No more attributes can be added after Fingerprint - it must  be
+             * the last attribute. */
+            result = STUN_RESULT_INVALID_ATTRIBUTE_ORDER;
+        }
+        else if( ( ( pCtx->attributeFlag & STUN_FLAG_INTEGRITY_ATTRIBUTE ) != 0 ) &&
+                 ( attributeType != STUN_ATTRIBUTE_TYPE_FINGERPRINT ) )
+        {
+            /* No attribute other than fingerprint can be added after Integrity
+             * attribute. */
+            result = STUN_RESULT_INVALID_ATTRIBUTE_ORDER;
+        }
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        /* Update flags. */
+        if( attributeType == STUN_ATTRIBUTE_TYPE_FINGERPRINT )
+        {
+            pCtx->attributeFlag |= STUN_FLAG_FINGERPRINT_ATTRIBUTE;
+        }
+        if( attributeType == STUN_ATTRIBUTE_TYPE_MESSAGE_INTEGRITY )
+        {
+            pCtx->attributeFlag |= STUN_FLAG_INTEGRITY_ATTRIBUTE;
+        }
+
+        if( pCtx->pStart != NULL )
+        {
+            /* Write Attribute type, length and value. */
+            STUN_WRITE_UINT16( (uint8_t *) &( pCtx->pStart[ pCtx->currentIndex ] ),
+                                attributeType );
+
+            STUN_WRITE_UINT16( (uint8_t *) &( pCtx->pStart[ pCtx->currentIndex + STUN_ATTRIBUTE_HEADER_LENGTH_OFFSET ] ),
+                                attributeValueBufferLength );
+
+            memcpy( ( void * ) &( pCtx->pStart[ pCtx->currentIndex + STUN_ATTRIBUTE_HEADER_VALUE_OFFSET ] ),
+                    pAttributeValueBuffer,
+                    attributeValueBufferLength );
+
+            /* Zero out the padded bytes. */
+            if( attributeValueLengthPadded > attributeValueBufferLength )
+            {
+                memset( ( void * ) &( pCtx->pStart[ pCtx->currentIndex + STUN_ATTRIBUTE_TOTAL_LENGTH( attributeValueBufferLength ) ] ),
+                        0,
+                        attributeValueLengthPadded - attributeValueBufferLength );
+            }
+        }
+
+        pCtx->currentIndex += STUN_ATTRIBUTE_TOTAL_LENGTH( attributeValueLengthPadded );
+    }
+
+    return result;
+}
+/*-----------------------------------------------------------*/
+
+static StunResult_t AddAttributeFlag( StunContext_t * pCtx,
+                                      StunAttributeType_t attributeType )
+{
+    StunResult_t result = STUN_RESULT_OK;
```

Comment:
L115 to L138 seems to be similar in other function. Can we have an helper function  for these tests or is it overkill or they are not all same (I'm lazy to check)